### PR TITLE
ast/parser: Implement parser for ast

### PR DESCRIPTION
**Pull request by coocood**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,7 @@
 /parser/parser.go
 /parser/scanner.go
+ast/parser/parser.go
```
Comment (qiuyesuifeng) [inline .gitignore]: /ast/parser/parser.go
/ast/parser/scanner.go

File: Makefile [hunk 2]
```diff
@@ -44,6 +44,24 @@ parser:
 
 	golex -o parser/scanner.go parser/scanner.l
 
+ast-parser:
+	a=`mktemp temp.XXXXXX`; \
```
Comment (qiuyesuifeng) [inline Makefile]: temp.ast.XXXXX

Comment (qiuyesuifeng) [inline Makefile]: ignore, perhaps needless.

File: ast/dml.go [hunk 3]
```diff
@@ -220,15 +307,15 @@ func (sn *SelectStmt) Accept(v Visitor) (Node, bool) {
...
 
-	for i, val := range sn.Unions {
-		node, ok := val.Accept(v)
+	if sn.Limit != nil {
+		node, ok := sn.Limit.Accept(v)
 		if !ok {
 			return sn, false
 		}
-		sn.Unions[i] = node.(*Union)
+		sn.Limit = node.(*Limit)
```
Comment (qiuyesuifeng) [inline ast/dml.go]: How about origin Unions ?

File: ast/dml.go [hunk 4]
```diff
@@ -23,10 +23,12 @@ var (
 	_ DMLNode = &UpdateStmt{}
 	_ DMLNode = &SelectStmt{}
 	_ Node    = &Join{}
-	_ Node    = &Union{}
-	_ Node    = &TableRef{}
+	_ Node    = &TableName{}
 	_ Node    = &TableSource{}
 	_ Node    = &Assignment{}
+	_ Node    = &Limit{}
+	_ Node    = &WildCardField{}
```
Comment (qiuyesuifeng) [inline ast/dml.go]: Wildcard?

File: ast/dml.go [hunk 5]
```diff
@@ -93,18 +104,18 @@ func (tr *TableRef) Accept(v Visitor) (Node, bool) {
...
 
-	// Source is the source of the data, can be a TableRef,
+	// Source is the source of the data, can be a TableName,
 	// a SubQuery, or a JoinNode.
 	Source Node
 
-	// Name is the alias name of the table source.
-	Name string
+	// Alias is the alias name of the table source.
+	Alias string
```
Comment (qiuyesuifeng) [inline ast/dml.go]: How about AsName?

File: ast/dml.go [hunk 6]
```diff
@@ -114,24 +125,25 @@ func (ts *TableSource) Accept(v Visitor) (Node, bool) {
...
 }
 
-// Union represents union select statement.
-type Union struct {
+// UnionClause represents a single "UNION SELECT ..." or "UNION (SELECT ...)" clause.
+type UnionClause struct {
 	node
 
-	Select *SelectStmt
+	Distinct bool
```
Comment (qiuyesuifeng) [inline ast/dml.go]: Why add a Distinct variable?

File: ast/dml.go [hunk 7]
```diff
@@ -316,10 +442,11 @@ func (in *InsertStmt) Accept(v Visitor) (Node, bool) {
 type DeleteStmt struct {
 	dmlNode
 
-	Tables      []*TableRef
+	TableRefs   *Join
```
Comment (qiuyesuifeng) [inline ast/dml.go]: A little confused why DeleteStmt has both TableRefs and Tables variables?

File: ast/dml.go (hunk 4 above)
Comment (coocood) [inline ast/dml.go]: I think WildCardField is easier to read.

File: ast/dml.go (hunk 7 above)
Comment (coocood) [inline ast/dml.go]: TableRefs is only used in multi-table mode.

File: ast/dml.go (hunk 6 above)
Comment (coocood) [inline ast/dml.go]: Union clause has a  distinct option.

File: ast/ddl.go [hunk 8]
```diff
@@ -35,12 +35,37 @@ var (
 	_ Node    = &AlterTableSpec{}
 )
 
+// FloatOpt is used for parsing floating-point type option from SQL.
+// TODO: add reference doc.
+type FloatOpt struct {
```
Comment (shenli) [inline ast/ddl.go]: Move those xxOpt into opt.go maybe better.

File: ast/ast.go [hunk 9]
```diff
@@ -70,11 +71,31 @@ type DMLNode interface {
...
+// ResultField is computed from a parsed select statement.
+type ResultField struct {
+	Column       *model.ColumnInfo
+	ColumnAsName model.CIStr
+	Table        *model.TableInfo
+	TableAsName  model.CIStr
+	DBName       model.CIStr
+}
+
+// ResultSetNode has result fields
```
Comment (shenli) [inline ast/ast.go]: Need more comments.

**Comment by shenli** [comment]
LGTM

**Comment by qiuyesuifeng** [comment]
LGTM

**Pull request merged**